File: C++/2231_분해합.cpp
#include <iostream>
#include <cmath>
#include <algorithm>
using namespace std;

int getDigit(int N);

int main() {
	int N;
	cin >> N;
	
	int digit = getDigit(N); // 주어진 수의 자릿수
	
	
	int con = 0; // 주어진 수가 분해합인 수 중 가장 작은 수
	int s = max(1,(int)pow(10,digit-2)); // 시작 수(주어진 수가 1자리 수라면 한자리 작은 수부터 볼 수 없음)
	for(int i = s; i < N; i++){
		int num = 0, tmp = i;
		while(tmp > 0){ // 해당 수의 각 자리 수를 잘라서 해당 숫자에 하나씩 더해 줌(분해합 구하는 방법)
			num += tmp % 10;
			tmp /= 10;
		}
		if(i+num == N){ // 해당 수의 분해합이 주어진 수와 같으면
			con = i; // 해당 수를 저장한 후
			break; // 루프 종료
		}
	}
	
	cout << con;
	
	return 0;
}

int getDigit(int N){ // 해당 수가 몇 자리 수인지 구하는 함수
	int digit = 1;
	
	N /= 10;
	while(N > 0){ // 한자리씩 줄이면서 해당 수가 몇 자리 수인지 알아내는 것
		digit++;
		N /= 10;
	}
	
	return digit;
}

File: Python/2231_분해합.py
import sys
from itertools import product

def getDigit(N): # 주어진 수가 몇 자릿수를 가지는지 구하는 함수
	digit = 0
	while N > 0:
		digit += 1
		N = int(N/10)
		
	return digit # 주어진 수의 자릿수

def getConstructor(N, digit):
	con = 0 # N의 가장 작은 생성자
	
	start = max(1,10**(digit-2)) # 확인하기 시작할 수(한 자리수 수가 주어졌을 때는 한 자리 작은 수를 확인할 수 없음)
	for num in range(start,N):
		tmp = 0
		for i in str(num): # 분해합 구하는 과정(해당 수 + 각 자리수의 수)
			tmp += int(i)
		if (tmp + num) == N: # 해당 생성자의 분해합이 주어진 수라면
			con = num # 생성자 저장 후 return할 것
			break

	return con
	

if __name__ == '__main__':
	N = int(sys.stdin.readline())
	
	digit = getDigit(N)
	print(getConstructor(N, digit))
